### ISSUE TITLE
alternative hybrid relationship loading strategy 

### ISSUE THREAD

**Issue by BrannJoly** [mikro-orm/mikro-orm#6129]
**Is your feature request related to a problem? Please describe.**
Right now we have [two relationship loading strategies ](https://mikro-orm.io/docs/loading-strategies) : joined and select-in

I was a happy user of "joined" (the default option for postgress), but lately I started noticing my backend was extremely slow.

I think the issue is related to postgress'[ joined collapse limit](https://www.postgresql.org/docs/current/runtime-config-query.html#GUC-JOIN-COLLAPSE-LIMIT). TLDR; if too many tables are joined, even if the final result is only a handfull of rows, even if all the tables are properly indexed, the query will be painfully slow.

I removed a few relationship to test my theory, and indeed it was fast again.

I tried switching to select-in, and this solves the issue, but creates another one : the requests look somewhat like select * from my_child_table where id in (.... a potentially very long list of all the parent entity id we want).
This long list of ids make this solution suboptimal

**Describe the solution you'd like**
add a new strategy "select-in-joined" that will execute one request per child entity (similar to the select-in strategy), but which will join on the parent entity rather than transferring all the invidivual ids (similar to the join strategy)

alternatively, update the select-in strategy so that it joins on the parents entity tables above a certain number of items (eg 100 by default)

**Comment by B4nan** [comment]
I am planning something similar, joining things only for to-one relations, and using separate selects for to-many relations. This way the cartesian products won't explode. Such a strategy would be the default most likely in v7.

I don't think a huge where condition is the problem here, if you think so, please provide a complete repro that proves that, otherwise I will go with the approach highlighted above.

**Comment by BrannJoly** [comment]
> I am planning something similar, joining things only for to-one relations, and using separate selects for to-many relations. This way the cartesian products won't explode. Such a strategy would be the default most likely in v7.

I think this would help a lot with my issue! Looking forward to V7, and thank you for the amazing work you're doing on mikro ORM !!

> I don't think a huge where condition is the problem here, if you think so, please provide a complete repro that proves that, otherwise I will go with the approach highlighted above.

I think there are two different issues:
1/ 'joined' doesn't work for me because postgress is notoriously bad at handling joins over too many tables (contrary to say MS SQLServer). That's why I like the select-in option, which avoids this issue entirely


2/ 'select-in' doesnt work either:

Let's say I have a table p (as parent entity) with 10.000 rows, and many tables c1..cn (as in child entity) with a FK on table p

If I want to load every single row, mikro orm will do this:
```
select * from p
select c1.* from c1 where p_id in (... a list of 10.000 ids)
.....
select cn.* from cn where p_id in (... a list of 10.000 ids)
```

We can't join all of thoses queries because that would be too many for postgress, but we coud at least do
```
select * from p
select c1.* from c1 inner join p on c1.p_id=p.id
....
select cn.* from cn inner join p on cn.p_id=p.id
```

This would be way more efficient, but it doesnt always work, eg:
```
select * from p where id in  (a very big and expensive subquery that returns only one id)
```
here, mikro orm current approach of passing the id(s) to each query seems totally warranted

That's why I think we would need a way to control on a query per query basis the strategy to use (or maybe this could be automated in mikro orm, I don't know!)

There were many assertions in my post, and it's unclear to me which one you would see a repro for!

**Comment by B4nan** [comment]
I see, I guess we could just batch those select calls, just like we do it with the insert/update queries from flush. We can do that easily now.

**Comment by BrannJoly** [comment]
> I see, I guess we could just batch those select calls, just like we do it with the insert/update queries from flush. We can do that easily now.

I think that would help a lot to make the select-in strategy more viable! I'm looking forward to testing that when it's available.

**Comment by B4nan** [comment]
I did a quick test and it looks quite fast on my end, a query with `where in 10000 ids` takes around 20ms locally with postgres. I guess things can go wrong with more complex scenarios, but without a repro its hard to optimize things.

**Comment by BrannJoly** [comment]
I did a quick example showing how much the "joined" strategy is broken using postgress when the number of collection grows. The repo is [here](https://github.com/BrannJoly/mikro-orm-bug-1)

Here is what I do:
I create a parent entity and 20 child entites (c1...c20). The parent entity has a collection of each.
I insert 10.000 parent entities, and 100 of each child entities in the db.
then for each parent, I randomly add to their c1 and c2 collections 10 childs. I don't touch the others c3 to c20 collections which remain completely empty

I then timed those three requests 
``` javascript
await orm.em.find(Parent, {}, {strategy: 'select-in', populate: ['c1s', 'c2s', 'c3s', 'c4s', 'c5s', 'c6s', 'c7s', 'c8s', 'c9s', 'c10s', 'c11s', 'c12s', 'c13s', 'c14s', 'c15s', 'c16s', 'c17s', 'c18s', 'c19s', 'c20s'] });
// 3915ms

await orm.em.find(Parent, {}, {strategy: 'joined', populate: ['c1s', 'c2s', 'c3s', 'c4s', 'c5s', 'c6s', 'c7s', 'c8s', 'c9s', 'c10s', 'c11s', 'c12s', 'c13s', 'c14s', 'c15s', 'c16s', 'c17s', 'c18s', 'c19s', 'c20s'] });
 // 38111ms 

  await orm.em.find(Parent, {}, {strategy: 'joined', populate: ['c1s', 'c2s'] });
// 4433ms
```

as you can see, joining on 18 addtional **empty tables** turns a 4s request into a 38s one ...

**Comment by B4nan** [comment]
I wanted to see a repro where the select in strategy is slow, joining to-many relations is destined to be problematic due to the cartesian products, and your repro is using 20 M:N relations, those are literally the worse, since they require two joins for every property. This is kinda destined to be slow, and feels like a very artificial thing to do. If someone is doing this in their apps, they must be crazy 🙃 

This is not really what you described as the culprit earlier, you said the problem is a huge `where in (...)` condition and I am not really sure that is the case.

**Comment by BrannJoly** [comment]
> I did a quick test and it looks quite fast on my end, a query with `where in 10000 ids` takes around 20ms locally with postgres. I guess things can go wrong with more complex scenarios, but without a repro its hard to optimize things.

yes it's quite fast locally with a trivial exemple.
Here is my real world scenario:
- the backend and the db are not on the same server

- i'm using uuids, not integers, so the primary keys are way bigger

- i'm using pagination_tokens, so I typically want to select the first n 1000 entities with a token>x. 

Calling our tables p (parent) and c (child), here is what Mirko orm does with select-in:
``` sql
select * from p where pagination_token>x limit 1000
select * from c inner join p_c_jointable where p_c_jointable.parent_id in (... the big list)
```

And here what I think would make way more sense (maybe not as default, but at least to have the option)
``` sql
select * from p where pagination_token>x limit 1000

select c.* from p 
inner join p_c_jointable on p_c_jointable.p_id=p.id
inner join c on p_c_jointable.c_id = c.id
where pagination_token>x limit 1000

```

**Comment by BrannJoly** [comment]
> I wanted to see a repro where the select in strategy is slow, joining to-many relations is destined to be problematic due to the cartesian products, and your repro is using 20 M:N relations, those are literally the worse, since they require two joins for every property. This is kinda destined to be slow, and feels like a very artificial thing to do. If someone is doing this in their apps, they must be crazy 🙃

you are right, I digress...

That being said, in my example I was comparing joining on 2 tables, and joining on those same 2 tables + 18 **_empty tables_** . This does absolutely nothing to the cartesian product, does it? the number of rows returned is exactly the same, it is just that postgress gets absurdely inefficient when handling too many joins! Hence the need for a finer grain control on the strategy to use. 

For example, maybe you could automatically fall back to the select-in strategy if the request is going to join too many tables on postgres

In conclusion, I will use select-in for everything until the new v7 hybrid strategy is available !

Thank you for your time!
Brann

**Comment by B4nan** [comment]
I wonder what hw you are using, I checked your repro and it behaves twice as fast on my end (M1 pro) :]

> 18 empty tables

Nope, they are not empty. The joined query you say is so slow returns 177786 results (and takes 1700ms on my end). The whole problem with that approach is not the query itself, but the hydration, caused by the huge cartesian product. You can see the numbers at the end of the queries. And that is exactly what could be solved with the approach I mentioned above.

> [took 1702 ms, 177786 results]

As opposed to the select in queries, those, even when containing huge `where in` clause, often take less than 10 ms when they don't return any results. So I still have doubts about this being a problem (and would still like to see a repro if you still believe in your claim).

> [took 8 ms, 0 results]

> And here what I think would make way more sense (maybe not as default, but at least to have the option)

Also regarding this suggestion, you cannot use the limit clause like this, again, because of the cartesian product.
